### PR TITLE
Update ServerCommon packages and start emitting issues from extract and validate

### PR DESCRIPTION
**Pull request by joelverhagen**
Complete https://github.com/NuGet/NuGetGallery/issues/5256.
Progress on https://github.com/NuGet/Engineering/issues/785.
Depends on https://github.com/NuGet/ServerCommon/pull/114.

This updates to new ServerCommon packages to allow extract and validate to persist its own issues. I will switch to a real package version when it is merged to master and published.

### CODE REVIEW

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidatorResult.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+using System;
+using System.Collections.Generic;
+using NuGet.Services.Validation;
+
+namespace NuGet.Jobs.Validation.PackageSigning.ExtractAndValidateSignature
+{
+    public class SignatureValidatorResult
+    {
+        public SignatureValidatorResult(ValidationStatus state, IReadOnlyList<IValidationIssue> issues)
```
Comment (loic-sharma) [inline src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidatorResult.cs]: >SignatureValidatorResult [](start = 15, length = 24)

In the message handler you only persist issues if the state is terminal. Thus, I think this constructor should throw if issues are provided but the state isn't terminal.

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs [hunk 2]
```diff
@@ -131,11 +138,15 @@ public class SignatureValidator : ISignatureValidator
...
-                await RejectAsync(validation, message);
-                return;
+                    errorsForLogs,
+                    warningsForLogs);
+
+                return await RejectAsync(
+                    packageKey,
+                    message,
+                    errorIssues
+                        .Select(x => new ClientSigningVerificationFailure(x.Code.ToString(), x.Message))
```
Comment (loic-sharma) [inline src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs]: >Code [](start = 76, length = 4)

I'm assuming that `x.Code` is an enum here? If not, should there be some sort of null check? Sorry, I'm too lazy to clone your repo for intellisense :P

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs [hunk 3]
```diff
@@ -158,26 +169,32 @@ private HashSet<string> GetThumbprints(IEnumerable<Signature> signatures)
...
         {
             await _packageSigningStateService.SetPackageSigningState(
-                validation.PackageKey,
+                packageKey,
                 message.PackageId,
                 message.PackageVersion,
                 status);
 
-            validation.State = ValidationStatus.Succeeded;
+            return new SignatureValidatorResult(ValidationStatus.Succeeded, new IValidationIssue[0]);
```
Comment (loic-sharma) [inline src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs]: >SignatureValidatorResult [](start = 23, length = 24)

Maybe the issues parameter should be optional?

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs (hunk 2 above)
Comment (joelverhagen) [reply src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs]: Yes, `x.Code` is an enum.

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs (hunk 3 above)
Comment (joelverhagen) [reply src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidator.cs]: Sure, sounds fine.

File: tests/Validation.PackageSigning.ExtractAndValidateSignature.Tests/SignatureValidationMessageHandlerFacts.cs [hunk 4]
```diff
@@ -142,53 +146,51 @@ public HandleAsync()
...
-
-            [Theory]
-            [InlineData(ValidationStatus.NotStarted)]
-            [InlineData(ValidationStatus.Incomplete)]
-            public async Task RejectsNonTerminalState(ValidationStatus state)
+            [InlineData(ValidationStatus.NotStarted, false)]
+            [InlineData(ValidationStatus.Incomplete, false)]
+            [InlineData(ValidationStatus.Failed, true)]
+            [InlineData(ValidationStatus.Succeeded, true)]
+            public async Task SavesStateAndMaybeIssues(ValidationStatus state, bool isTerminal)
```
Comment (chenriksson) [inline tests/Validation.PackageSigning.ExtractAndValidateSignature.Tests/SignatureValidationMessageHandlerFacts.cs]: `Maybe` is yes when terminal state? Maybe split into two tests: `SavesStateWithIssuesWhenTerminal` and `SavesStateWithoutIssuesWhenNonTerminal`?

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidationMessageHandler.cs [hunk 5]
```diff
@@ -112,7 +113,28 @@ public class SignatureValidationMessageHandler
...
+                    message,
+                    cancellationToken);
+
+                validation.State = result.State;
+
+                // Save any issues if the resulting state is terminal.
+                if (validation.State == ValidationStatus.Failed
+                    || validation.State == ValidationStatus.Succeeded)
+                {
+                    validation.ValidatorIssues = validation.ValidatorIssues ?? new List<ValidatorIssue>();
```
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidationMessageHandler.cs]: We named `Issues` to eventually handle both errors and warnings. Eventually, a successful validation could have warnings.

File: tests/Validation.PackageSigning.ExtractAndValidateSignature.Tests/SignatureValidationMessageHandlerFacts.cs (hunk 4 above)
Comment (joelverhagen) [reply tests/Validation.PackageSigning.ExtractAndValidateSignature.Tests/SignatureValidationMessageHandlerFacts.cs]: Sure, had 'em split before but merged to simplicity. I'll split again.

File: src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidatorResult.cs (hunk 1 above)
Comment (joelverhagen) [reply src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidatorResult.cs]: Sure, sounds fine.

Comment (joelverhagen) [reply src/Validation.PackageSigning.ExtractAndValidateSignature/SignatureValidatorResult.cs]: Fixed.

**Pull request merged**